### ISSUE TITLE
await on IMessageBroker.Receive<T> ?

### ISSUE THREAD

**Issue by slimshader** [neuecc/UniRx#375]
in MonoBehaviour 1 I have a call:

```
await messageBroker.Receive<Message>();
Debug.Log("Received");
```

while in the other I have:
```
messageBroker.Publish<Message>();
Debug.Log("Published");
```

my problem is that program seems to stop at await stage, no "Publish" or "Received" is printed.

publishing an event is definately happening after await instruction

**Comment by slimshader** [comment]
It seems that it happens with Subject also meaning awaiting it never continues even the OnNext was called on it elsewhere

**Comment by slimshader** [comment]
I understand the problem now. 'await' on Observable waits for the last result before completion, not the first one. The solution is to add FirstAsync() operator to the Rx library and await call like so
```
await messageBroker.Receive<Message>().FirstAsync();
```

**Comment by neuecc** [comment]
Yes, Rx's standard await if last.
FirstAsync or convert to the UniTask, may works fine.

**Comment by neitron** [comment]
> Yes, Rx's standard await is last. FirstAsync or convert to the UniTask, may works fine.

I found out that `await someSubject.ToUniTask();` works like a charm. 
But `await MessageBroker.Default.Receive<SomeMessage>().ToUniTask()` doesn't. 
@neuecc Can u help me here and bring some light on it?

**Comment by slimshader** [comment]
> > Yes, Rx's standard await is last. FirstAsync or convert to the UniTask, may works fine.
> 
> I found out that `await someSubject.ToUniTask();` works like a charm. But `await MessageBroker.Default.Receive<SomeMessage>().ToUniTask()` doesn't. @neuecc Can u help me here and bring some light on it?


ToUniTask() takes bool parameter of whether it should be awaiting First of last element for IObservable().

**Comment by neitron** [comment]
> 

`await MessageBroker.Default.Receive<SomeMessage>().Take(1).ToUniTask()` WORKS!

**Comment by slimshader** [comment]
> 
> `await MessageBroker.Default.Receive<SomeMessage>().Take(1).ToUniTask()` WORKS!

Yes, and is effectively the same as Receive().First().ToUniTask()
